### ISSUE TITLE
[BUG] 4.9.0 crashing with Avalonia on Linux

### ISSUE THREAD

**Issue by Laurnz** [emgucv/emgucv#955]
I am using EmguCV with the UI framework Avalonia on Ubuntu and the application crashes during startup of the UI. Since this is only happening on version 4.9.0.5494, I suspect EmguCV to be the culprit, as this was not happening before.
This happens with the Ubuntu 22.04 NuGet package, or also on Ubuntu 24.04 with a self build runtime package.

This does not happen on Windows, when not using Avalonia, or when using an earlier version of EmguCV.

**OS / Platform**
Ubuntu 22.04, Ubuntu 24.04

**.Net version**
.Net 8, x86_64

**Emgu CV package used**
Emgu.CV.runtime.ubuntu-x64 nuget version 4.9.0.5494

**To Reproduce**
Here is a sample project to run: https://github.com/Laurnz/AvaloniaEmguCV
This is an official Avalonia sample project, where I additionally instantiated a EmguCV.Mat in the Program.cs.

**Expected behavior**
The app crashes before the UI is shown without throwing a catchable exception.

**Additional Information**
From what I can see in the coredump, it seems to be crashing while calling another native library. I am not skilled enough to debug this problem further.
```
Child SP               IP Call Site
00007FFF7ABB12B0 0000734ecbd5c42f [InlinedCallFrame: 00007fff7abb12b0] HarfBuzzSharp.HarfBuzzApi.hb_font_create(IntPtr)
00007FFF7ABB12B0 0000734e4e25c199 [InlinedCallFrame: 00007fff7abb12b0] HarfBuzzSharp.HarfBuzzApi.hb_font_create(IntPtr)
00007FFF7ABB12A0 0000734E4E25C199 HarfBuzzSharp.Font..ctor(HarfBuzzSharp.Face)
00007FFF7ABB1360 0000734E4E25AE84 Avalonia.Skia.GlyphTypefaceImpl..ctor(SkiaSharp.SKTypeface, Avalonia.Media.FontSimulations)
00007FFF7ABB1470 0000734E4E257341 Avalonia.Skia.FontManagerImpl.TryCreateGlyphTypeface(System.IO.Stream, Avalonia.Media.FontSimulations, Avalonia.Media.IGlyphTypeface ByRef)
00007FFF7ABB14B0 0000734E4E25338D Avalonia.Media.Fonts.EmbeddedFontCollection.Initialize(Avalonia.Platform.IFontManagerImpl)
00007FFF7ABB1690 0000734E4E252AAF Avalonia.Media.FontManager.AddFontCollection(Avalonia.Media.Fonts.IFontCollection)
00007FFF7ABB1710 0000734E4E253035 Avalonia.AppBuilderExtension+<>c.<WithInterFont>b__0_0(Avalonia.Media.FontManager)
00007FFF7ABB1740 0000734E4E2512D0 Avalonia.AppBuilder+<>c__DisplayClass71_0.<ConfigureFonts>b__0(Avalonia.AppBuilder)
00007FFF7ABB1780 0000734E4D216E66 ILStubClass.IL_STUB_MulticastDelegate_Invoke(System.__Canon)
00007FFF7ABB17B0 0000734E4D232344 Avalonia.AppBuilder.SetupUnsafe()
00007FFF7ABB1890 0000734E4D231FE0 Avalonia.AppBuilder.Setup()
00007FFF7ABB1900 0000734E4D231DDD Avalonia.AppBuilder.SetupWithLifetime(Avalonia.Controls.ApplicationLifetimes.IApplicationLifetime)
00007FFF7ABB1920 0000734E4D217082 Avalonia.ClassicDesktopStyleApplicationLifetimeExtensions.StartWithClassicDesktopLifetime(Avalonia.AppBuilder, System.String[], System.Action`1<Avalonia.Controls.ApplicationLifetimes.IClassicDesktopStyleApplicationLifetime>)
00007FFF7ABB1950 0000734E4D21198C AvaloniaTest.Program.Main(System.String[])
```

**Comment by Laurnz** [comment]
I don't know if #918 is related, or if it could be the fault of Avalonia/SkiaSharp. But it is really specific to the new version of EmguCV 4.9.0.5494

**Comment by emgucv** [comment]
The error log doesn't contains any calls to the Emgu library before it crashes. There are only calls to the Avalonia and HarbuzzSharp libraries available in the log. 

You mention the error is specific to the new version of Emgu CV 4.9.0.5494, does that means if you use an older version of Emgu CV library and it works without error?

**Comment by Laurnz** [comment]
Yes, the program does not crash directly in the EmguCV part. The program calls EmguCV first and then crashes when initializing the UI framework later. And it is not happening with older versions of EmguCV

We have used EmguCV together with the UI framework for a long time, and with different versions. This crash is only happening when using the new 4.9.0.5494 version.

This is why I opened the issue here and not in the Avalonia/SkiaSharp repository, as the regression was coupled to the EmguCV version number, even though not crashing directly there.

**Comment by emgucv** [comment]
What is the exact version of Emgu CV that works without any issues? When it works, does it run on Ubuntu 22.04 (or 24.04)?

FYI, 4.9.0 release is built on Ubuntu 22.04, and is only expected to run on Ubuntu 22.04. We do not support Ubuntu 24.04 for the 4.9.0 release.

**Comment by Laurnz** [comment]
I specifically tested 4.8.1.5350 and also used lower ones over the years. I always tried Ubuntu 22.04.
If you do not manage to get my sample repo to execute after changing EmguCV version to 4.8.1.5350, then I must have screwed up something.

I just mentioned Ubuntu 24.04, because I self build the binaries for that Ubuntu version and originally found the issue on that distro, but then validated it also on 22.04. Sorry, mentioning Ubuntu 24.04 was confusing.

**Comment by emgucv** [comment]
Nuget package tested to work on Ubuntu 22.04 using the following steps:

1. Create a new project folder
2. Create a new console project 
`dotnet new console`
3.  Add the Emgu CV 4.9 ubuntu runtime:
`dotnet add package Emgu.CV.runtime.ubuntu-x64 --version 4.9.0.5494`
4. Replace Program.cs file with this file:
https://github.com/emgucv/emgucv/blob/18a01c223b37e975ac3af59de45464425a13cec5/Emgu.CV.Example/Docker/Program.cs#L1
5. Run the program:
`dotnet run`

![image](https://github.com/user-attachments/assets/0269f361-f31a-4302-a8aa-8ea8e8e09528)

**Comment by Laurnz** [comment]
As mentioned, running just EmguCV works fine. The problem is when other this other library is also calling native code.

I opened the issue here, because the version change introduced the exception. But maybe it makes more sense opening an issue in the SkiaSharp repository. So I will do that.

**Comment by emgucv** [comment]
Root cause found — this is the same defect that was diagnosed in #1027 and #975, and it is fixed in e5072ab3691259d6df5fc9323735d414c767dd73. Your diagnosis was correct all along: Emgu CV was breaking *other* libraries' native code.

On Linux, Emgu CV loaded `libcvextern.so` with `dlopen(..., RTLD_NOW | RTLD_GLOBAL)`. `RTLD_GLOBAL` promotes the symbols of the library — and of its entire dependency subtree — into the process's global symbol scope, where they interpose the internal PLT calls of native libraries loaded afterwards. In #1027 we demonstrated this concretely with `LD_DEBUG=bindings`: EF Core's bundled SQLite had 927 of its 934 internal `sqlite3_*` calls hijacked to a different library from Emgu's dependency chain. Avalonia's SkiaSharp native library is vulnerable to exactly the same interposition (Skia bundles its own copies of common libraries such as freetype and harfbuzz, whose symbols also travel in Emgu's dependency tree), which is why the UI dies at startup only when Emgu's native library is loaded first — and why the failure appeared with the 4.9.0 release, whose Ubuntu build changed the bundled dependency set.

The fix loads native libraries with the default `RTLD_LOCAL` scope (nothing in Emgu required global visibility), which makes every later-loaded library resolve its own symbols internally. It ships with the next release.

**Workaround for current versions:** initialize Avalonia/Skia *before* the first Emgu CV call (e.g. touch a SkiaSharp type before constructing any `Mat`) — a library whose symbols are bound before Emgu loads is immune to the interposition, which also explains the load-order-dependent behavior you observed. Keeping this open until the fixed packages are published.
